### DISCUSSION TITLE
it  is 45   that reading  value of gd->arch.boot_hart .,why?

### DISCUSSION THREAD

**Discussion by yeliangchen2** [polarfire-soc/polarfire-soc-documentation#158] [HSS and U-Boot] [answered]
to run normally, the value of gd->arch.boot_hart  should be 1~4 .but I read it ,and it is 45.why?

boot_f  stage :the value of gd->arch.boot_hart is 45.
boot_r  stage :the value of gd->arch.boot_hart is also 45.

env: 
cpu :mpfs025T
hss load payload.bin addr :0x80200000
uboot addr to run: 0x10 00200000
mem max size :256M,  and 192M for uboot (0x10 00000000~0x10 0C000000)

yaml config:
![image](https://user-images.githubusercontent.com/16462894/181666548-59f7f971-a13e-4e67-a211-d0b07f2b2395.png)

question: who changed the value of gd->arch.boot_hart ?

**Comment by hughbreslin** [comment]
Hi @yeliangchen2 how are you seeing this value being read? We're going to need more detail on this. Is it happening all of the time and consistently? It seems like memory corruption, what DDR is this? Can you run a `memtest` from the HSS and verify it passes?

**Reply by yeliangchen2** [reply]
I runned memtest ,and it passed.  
In the "boot_init_f" and "boot_init_r" stage,I use printf function to output value of "gd->arch.boot_hart",as this "pritf("gd->arch.boot_hart=%d \n", gd->arch.boot_hart)".
as follows:
define my function:
![image](https://user-images.githubusercontent.com/16462894/181865002-f7b47dad-a3b4-4585-b0a8-a1e4ecd38648.png)
call my function in the boot_init_f,and the value of gd->arch.boot_hart is 45 :
![image](https://user-images.githubusercontent.com/16462894/181865020-f6ca4140-2861-4769-a932-42eb57e55190.png)
call my function in the boot_init_r,and the value of gd->arch.boot_hart is 45 :
![image](https://user-images.githubusercontent.com/16462894/181865097-56bc176c-3668-4d00-a0b9-17db1b209bf9.png)

if i read the value of gd->arch.boot_hart is 45 ,and "riscv_timer" can not add to uclass, as a result  uboot is failed to run.
![image](https://user-images.githubusercontent.com/16462894/181865162-f327f79e-c05a-40db-b815-a35139829b65.png)

I read up the code of uboot ,find it getting hardid in start.s file

**Reply by hughbreslin** [reply]
Hmm ok, could you add an ecall to the start.s file where the hartid is set and then connect the debugger to verify it is correct at that point? You should then be able to set a watchpoint to see where / when it changes.

**Reply by vishakh-mchp** [reply]
Could you please elaborate how to add an ecall - this is something that we do not have experience doing.

**Reply by hughbreslin** [reply]
Sure, there are two instruction, `ebreak` and `ecall` the HAL has an example of `ebreak` [here](https://github.com/polarfire-soc/platform/blob/main/mpfs_hal/startup_gcc/mss_entry.S#L291) just add the instruction into start.s after the hartid has been read and the application will halt at that point

**Comment by xiyanni** [comment, marked as answer]
When our FAE migrated and built the HSS project by himself, this error disappeared on customer board.  Uboot works now.  Issue reported by customer maybe caused by wrong HSS configuration.
